stl/inc/span

```diff
@@ -0,0 +1,665 @@
+template <class _Cont, class _Ty>
+inline constexpr bool _Is_span_convertible_range_v =
+    is_convertible_v<remove_pointer_t<decltype(_STD data(_STD declval<_Cont&>()))> (*)[], _Ty (*)[]>;
+
+// VARIABLE TEMPLATE _Has_container_interface_v
+template <class, class = void>
+inline constexpr bool _Has_container_interface_v = false;
+
+template <class _Cont>
+inline constexpr bool _Has_container_interface_v<_Cont,
+    void_t<decltype(_STD data(_STD declval<_Cont&>())), decltype(_STD size(_STD declval<_Cont&>()))>> = true;
+
+// VARIABLE TEMPLATE _Is_span_compatible_container_v
+// clang-format off
+template <class _Cont, class _Ty>
+inline constexpr bool _Is_span_compatible_container_v =
+       !_Is_span_v<remove_const_t<_Cont>>
+    && !_Is_std_array_v<remove_const_t<_Cont>>
+    && !is_array_v<_Cont>
+    && _Is_span_convertible_range_v<_Cont, _Ty>
+    && _Has_container_interface_v<_Cont>;
+// clang-format on
+
+// [views.span]
+// CLASS TEMPLATE span
+template <class _Ty, size_t _Extent = dynamic_extent>
+class span : public _Span_extent_type<_Extent> {
+public:
+    using _Mybase                = _Span_extent_type<_Extent>;
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using index_type             = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr index_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+    constexpr span() noexcept = default;
+
+    constexpr span(pointer _Ptr, index_type _Count) noexcept // strengthened
+        : _Mybase(_Count), _Mydata(_Ptr) {}
+
+    constexpr span(pointer _First, pointer _Last) noexcept // strengthened
+        : _Mybase(static_cast<index_type>(_Last - _First)), _Mydata(_First) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_First <= _Last, "span construction from pointers only valid for first <= last");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    template <size_t _Size, enable_if_t<_Extent == dynamic_extent || _Extent == _Size, int> = 0>
+    constexpr span(element_type (&_Arr)[_Size]) noexcept : _Mybase(_Size), _Mydata(_Arr) {}
+
+    template <size_t _Size, enable_if_t<_Extent == dynamic_extent || _Extent == _Size, int> = 0>
+    constexpr span(array<value_type, _Size>& _Arr) noexcept : _Mybase(_Size), _Mydata(_Arr.data()) {}
+
+    template <size_t _Size, enable_if_t<_Extent == dynamic_extent || _Extent == _Size, int> = 0>
+    constexpr span(const array<value_type, _Size>& _Arr) noexcept : _Mybase(_Size), _Mydata(_Arr.data()) {}

```

Comment:
Where is the https://eel.is/c++draft/span.cons#11.2 constraint? It's been a while since I thought deeply about these constraints, but I believe that this one prevents `span<T>` from being constructible from `const array<T, N>`.

Reply:
I am having trouble to correctly implement that constraint. The problem I face is that for constructors it is not possible to access the function parameters for SFINAE, which I would normally do in the return type.

Consequently I have no Idea how to SFINAE on `data(_arr)` when there is no `_arr` available.

On the other hand I could hard code the return type of `data(_arr)` and use that. But that seems meh too.

Reply:
You can use `_STD declval<array<value_type, _Size>&>()` to get an "imaginary" `array` lvalue.

Reply:
This constraint is trivially satisfied for 377: `data(_Arr)` returns `element_type*` and `element_type(*)[]` is certainly convertible to `element_type(*)[]`. It's also trivially satisfied for 380: `data(_Arr)` returns `value_type*`, and `value_type(*)[]` is certainly convertible to `element_type(*)[]` since `value_type` is `remove_cv_t<element_type>`. So we need only concern ourselves with 383.

For 383, `data(_Arr)` returns `const value_type*`, so it suffices to verify `is_convertible_v<const value_type(*)[], element_type(*)[]>`, which is equivalent to `is_convertible_v<const value_type*, element_type*>` since (again) we know that `value_type` is `remove_cv_t<element_type>`.

TLDR: 383 needs `&& is_convertible_v<const value_type*, element_type*>` added to the first argument to `enable_if_t`.

Reply:
Done